components/docs-chef-io/content/habitat/origin-keys.md

```diff
@@ -0,0 +1,190 @@
++++
+title = "Origin Keys"
+
+date = 2020-10-12T13:59:46-07:00
+draft = false
+
+[menu]
+  [menu.habitat]
+    title = "Origin Keys"
+    identifier = "habitat/origins/origin-keys Origin Keys"
+    parent = "habitat/origins"
+    weight = 20
++++
+
+[\[edit on GitHub\]](https://github.com/habitat-sh/habitat/blob/master/components/docs-chef-io/content/habitat/origin-keys.md)
+
+Prerequisites:
+
+- [Get Chef Habitat]({{< relref "install-habitat.md" >}})
+- [Create a Chef Habitat Builder account]({{< relref "builder-account#builder-account" >}})
+- [Generate a personal access token]({{< relref "builder-profile#create-a-personal-access-token" >}})
+- [Create an origin]({{< relref "origins" >}}) or join an origin by [invitation]({{< relref "origin-rbac#manage-origin-membership-with-hab-origin-invitations" >}})
+
+When you create an origin, Chef Habitat Builder automatically generates _origin keys_.
+Origin key cryptography is asymmetric: it has a public origin key that you can distribute freely, and a private origin key that you should distribute only to users of the origin.
+All Chef Habitat Builder users with access to the origin can view the origin public key revisions in the origin key tab (Builder > Origin > Keys) and download the origin public key, but only users with the origin 'administrator' or 'owner' roles can view or download the origin private key, or change the origin key pair.
+
+| Keys Actions | Read-Only | Member | Maintainer | Administrator | Owner |
+|---------|-------|-------|-------|-------|-------|
+| View keys | Y | Y | Y | Y | Y |
+| Add/Update/Delete keys | N | N | N | Y | Y |
+
+Chef Habitat uses origin keys:
+
+- When you build an artifact in your local environment, Chef Habitat signs the artifact with a public key
+- When you upload an artifact to Chef Habitat Builder or Builder on-prem, Chef Habitat verifies that the artifact was signed with its public key
```

Comment:
This may be ambiguous. Is Habitat verifying "that the artifact was signed with its public key", or verifying, using the public key, that the artifact was signed?

(It should be the latter)

Reply:
First, let me tell you that your origin_keys.rs code is really easy to follow.

It should be the latter.